src/main/java/org/csanchez/jenkins/plugins/kubernetes/pod/retention/Reaper.java

```diff
@@ -169,4 +177,99 @@ private void closeWatch() {
+        @Override
+        public void onEvent(@NonNull Action action, @NonNull KubernetesSlave node, @NonNull Pod pod) {
+            if (action != Action.MODIFIED) {
+                return;
+            }
+            List<ContainerStatus> terminatedContainers = PodUtils.getTerminatedContainers(pod);
+            if (!terminatedContainers.isEmpty()) {
+                String ns = pod.getMetadata().getNamespace();
+                String name = pod.getMetadata().getName();
+                TaskListener runListener = node.getTemplate().getListener();
+                terminatedContainers.forEach(c -> {
+                    ContainerStateTerminated t = c.getState().getTerminated();
+                    LOGGER.info(() -> ns + "/" + name + " Container " + c.getName() + " was just terminated, so removing the corresponding Jenkins agent");
+                    runListener.getLogger().printf("%s/%s Container %s was terminated (Exit Code: %d, Reason: %s)%n", ns, name, c.getName(), t.getExitCode(), t.getReason());
+                });
+                Computer computer = node.toComputer();
+                if (computer != null) {
+                    computer.getExecutors().forEach(exec -> exec.interrupt());
+                }
+                try {
+                    node.terminate();
+                } catch (Exception x) {
+                    LOGGER.log(Level.WARNING, "failed to reap " + ns + "/" + name, x);
+                }
+            }
+        }
+    }
+
+    @Extension
+    public static class TerminateAgentOnImagePullBackOff extends Listener {
+
+        @Override
+        public void onEvent(@NonNull Action action, @NonNull KubernetesSlave node, @NonNull Pod pod) {
+            String ns = pod.getMetadata().getNamespace();
+            String name = pod.getMetadata().getName();
+            List<ContainerStatus> backOffContainers = PodUtils.getContainers(pod, cs -> {
+                ContainerStateWaiting waiting = cs.getState().getWaiting();
+                return waiting != null && waiting.getMessage() != null && waiting.getMessage().contains("Back-off pulling image");
+            });
+            if (backOffContainers.isEmpty()) {
+                return;
+            }
+            backOffContainers.forEach(cs -> {
+                TaskListener runListener = node.getTemplate().getListener();
+                runListener.error("Unable to pull Docker image \""+cs.getImage()+"\". Check if image tag name is spelled correctly.");
+            });
+            Queue q = Jenkins.get().getQueue();
+            String runUrl = pod.getMetadata().getAnnotations().get("runUrl");
+            for (Queue.Item item: q.getItems()) {
+                if (item.task.getUrl().equals(runUrl)) {
+                    q.cancel(item);
+                    break;
+                }
+            }
+            try {
+                node.terminate();
+            } catch (Exception x) {
```

Comment:
Better for the caller at https://github.com/jenkinsci/kubernetes-plugin/pull/772/files#diff-55ff61b56516ad2fa5dd36df80c3836cR155 to `catch Exception` for any listener